src/Data/Vec/Properties.agda

```diff
@@ -281,388 +270,383 @@ updateAt-cong i f≗g xs = updateAt-cong-relative i xs (f≗g (lookup xs i))
+  zipWith-zeroˡ : LeftZero _≡_ e f →
                   LeftZero _≡_ (replicate e) (zipWith {n = n} f)
   zipWith-zeroˡ zeˡ []       = refl
   zipWith-zeroˡ zeˡ (x ∷ xs) =
-    P.cong₂ _∷_ (zeˡ x) (zipWith-zeroˡ zeˡ xs)
+    cong₂ _∷_ (zeˡ x) (zipWith-zeroˡ zeˡ xs)
 
-  zipWith-zeroʳ : RightZero _≡_ e f → ∀ {n} →
+  zipWith-zeroʳ : RightZero _≡_ e f →
                   RightZero _≡_ (replicate e) (zipWith {n = n} f)
   zipWith-zeroʳ zeʳ []       = refl
   zipWith-zeroʳ zeʳ (x ∷ xs) =
-    P.cong₂ _∷_ (zeʳ x) (zipWith-zeroʳ zeʳ xs)
+    cong₂ _∷_ (zeʳ x) (zipWith-zeroʳ zeʳ xs)
+
+module _ {f : A → A → A} {e : A} {⁻¹ : A → A} where
 
-  zipWith-inverseˡ : ∀ {⁻¹} → LeftInverse _≡_ e ⁻¹ f → ∀ {n} →
+  zipWith-inverseˡ : LeftInverse _≡_ e ⁻¹ f →
                      LeftInverse _≡_ (replicate {n = n} e) (map ⁻¹) (zipWith f)
   zipWith-inverseˡ invˡ []       = refl
   zipWith-inverseˡ invˡ (x ∷ xs) =
-    P.cong₂ _∷_ (invˡ x) (zipWith-inverseˡ invˡ xs)
+    cong₂ _∷_ (invˡ x) (zipWith-inverseˡ invˡ xs)
 
-  zipWith-inverseʳ : ∀ {⁻¹} → RightInverse _≡_ e ⁻¹ f → ∀ {n} →
+  zipWith-inverseʳ : RightInverse _≡_ e ⁻¹ f →
                      RightInverse _≡_ (replicate {n = n} e) (map ⁻¹) (zipWith f)
   zipWith-inverseʳ invʳ []       = refl
   zipWith-inverseʳ invʳ (x ∷ xs) =
-    P.cong₂ _∷_ (invʳ x) (zipWith-inverseʳ invʳ xs)
+    cong₂ _∷_ (invʳ x) (zipWith-inverseʳ invʳ xs)
 
 module _ {f g : A → A → A} where
 
-  zipWith-distribˡ : _DistributesOverˡ_ _≡_ f g → ∀ {n} →
+  zipWith-distribˡ : _DistributesOverˡ_ _≡_ f g →
                      _DistributesOverˡ_ _≡_ (zipWith {n = n} f) (zipWith g)
   zipWith-distribˡ distribˡ []        []      []       = refl
   zipWith-distribˡ distribˡ (x ∷ xs) (y ∷ ys) (z ∷ zs) =
-    P.cong₂ _∷_ (distribˡ x y z) (zipWith-distribˡ distribˡ xs ys zs)
+    cong₂ _∷_ (distribˡ x y z) (zipWith-distribˡ distribˡ xs ys zs)
 
-  zipWith-distribʳ : _DistributesOverʳ_ _≡_ f g → ∀ {n} →
+  zipWith-distribʳ : _DistributesOverʳ_ _≡_ f g →
                      _DistributesOverʳ_ _≡_ (zipWith {n = n} f) (zipWith g)
   zipWith-distribʳ distribʳ []        []      []       = refl
   zipWith-distribʳ distribʳ (x ∷ xs) (y ∷ ys) (z ∷ zs) =
-    P.cong₂ _∷_ (distribʳ x y z) (zipWith-distribʳ distribʳ xs ys zs)
+    cong₂ _∷_ (distribʳ x y z) (zipWith-distribʳ distribʳ xs ys zs)
 
-  zipWith-absorbs : _Absorbs_ _≡_ f g → ∀ {n} →
+  zipWith-absorbs : _Absorbs_ _≡_ f g →
                    _Absorbs_ _≡_ (zipWith {n = n} f) (zipWith g)
   zipWith-absorbs abs []       []       = refl
   zipWith-absorbs abs (x ∷ xs) (y ∷ ys) =
-    P.cong₂ _∷_ (abs x y) (zipWith-absorbs abs xs ys)
+    cong₂ _∷_ (abs x y) (zipWith-absorbs abs xs ys)
 
 module _ {f : A → A → B} where
 
-  zipWith-comm : (∀ x y → f x y ≡ f y x) → ∀ {n}
-                 (xs ys : Vec A n) → zipWith f xs ys ≡ zipWith f ys xs
+  zipWith-comm : ∀ (comm : ∀ x y → f x y ≡ f y x) (xs ys : Vec A n) →
+                 zipWith f xs ys ≡ zipWith f ys xs
```

Comment:
It's kind of strange to be insisting on `f` being commutative as opposed to
the more general constraint `∀ x y → f x y ≡ g y x`.

Reply:
Generalised. At the time I was trying to lift algebraic structures over vectors, so I was thinking in more general terms.

Reply:
Makes sense. I guess now that we're not insisting on commutativity, we can also relax the type and have:
```agda
f : A → B → C
g : B → A → C
```

Reply:
Generalised further!